src/trace.c

```diff
@@ -1040,24 +1074,47 @@ trace_add_path (const void *parent,
 	if (! path_hash)
 		path_hash = NIH_MUST (nih_hash_string_new (NULL, 2500));
 
-	if (nih_hash_lookup (path_hash, pathname)) {
+	/* Try to insert a new path into hash table,
+	 * or abort in case if it already exists (processed).
+	 */
+	if (! maybe_insert_new_path (path_hash, pathname)) {
 		return 0;
-	} else {
-		NihListEntry *entry;
-
-		entry = NIH_MUST (nih_list_entry_new (path_hash));
-		entry->str = NIH_MUST (nih_strdup (entry, pathname));
-
-		nih_hash_add (path_hash, &entry->entry);
 	}
 
-	/* Make sure that we have an ordinary file
-	 * This avoids us opening a fifo or socket or symlink.
+	/* Start checking the file type to see if it can be
+	 * resolved down to some ordinary file.
 	 */
-	if ((lstat (pathname, &statbuf) < 0)
-	    || (S_ISLNK (statbuf.st_mode))
-	    || (! S_ISREG (statbuf.st_mode)))
+	if (lstat (pathname, &statbuf) < 0) {
 		return 0;
+	}
+
+	if (! S_ISREG (statbuf.st_mode)) {
```

Comment:
nit: or you can just do `if (S_ISLINK...` first and then this `! S_ISREG` path?

Reply:
Done, thank you for suggestion!